pkg/execution/pauses/manager.go

```diff
@@ -0,0 +1,140 @@
+package pauses
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/inngest/inngest/pkg/execution/state"
+	"github.com/inngest/inngest/pkg/logger"
+)
+
+var (
+	defaultFlushDelay = 10 * time.Second
+)
+
+// StateBufferer transforms a state.Manager into a state.Bufferer.
+func StateBufferer(rsm state.Manager) Bufferer {
+	return &redisAdapter{rsm}
+}
+
+// NewManager returns a new pause writer, writing pauses to a Valkey/Redis/MemoryDB
+// compatible buffer
+func NewManager(buf Bufferer, bs BlockStore) *manager {
+	return &manager{
+		buf: buf,
+		bs:  bs,
+		// XXX: should be able to override this
+		flushDelay: defaultFlushDelay,
+	}
+}
+
+type manager struct {
+	buf        Bufferer
+	bs         BlockStore
+	flushDelay time.Duration
+}
+
+func (m manager) ConsumePause(ctx context.Context, id uuid.UUID, data any) (state.ConsumePauseResult, error) {
+	// NOTE: There is a race condition when flushing blocks:  we may copy a pause
+	// into a block, then while writing the block to disk delete/consume a pause
+	// that is being written.  Unfortunately, in this case the metadata for a block
+	// isn't yet in the index. EG:
+	//
+	// 1. We read the buffer and add to a block
+	// 2. And while uploading the block
+	// 3. Kn parallel, we may delete one of the buffer’s pauses
+	//
+	// Unfortunately, we only write the block to indexes after uploads complete.
+	// This means that a pause may exist in a block but have been consumed.
+	//
+	// This is fine technically speaking:  consuming pauses is idempotent and leases
+	// each pause.
+	//
+	// However, in order to eventually compact we need to handle the “pause not found”
+	// case when consuming, and always re-delete the pause.  that’s no big deal, but
+	// not the best.
+	//
+	// In the future, we could add two block indexes:  pending, and stored.  this is a
+	// pain, though, because we may die when uploading pending blocks, and that requires
+	// a bit of thought to work around, so we’ll just go with double deletes for now,
+	// assuming this won’t happen a ton.  this can eb optimized later
```

Comment:
got it, could we add metrics to see how often this happens?
and also warning logs.